### PR TITLE
Add quantization on LLMs.

### PR DESCRIPTION
**Pull request by RogerRogerUSC**
I added quantization compression on the full connection layers of LLMs.

### CODE REVIEW

File: cezo_fl_main.py [hunk 1]
```diff
@@ -70,6 +71,7 @@ def prepare_settings_underseed(args, device):
         large_model = args.large_model
         model_name = SUPPORTED_LLM[large_model]
         model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch_dtype).to(device)
+        replace_penultimate_layer(model)
```
Comment (BichengYing) [inline cezo_fl_main.py:74]: please do not modify the cezo_fl main file. It should be place in a separate file.

File: cezo_fl_main.py [hunk 2]
```diff
@@ -98,6 +100,12 @@ def prepare_settings_underseed(args, device):
     return model, criterion, optimizer, grad_estimator, accuracy_func
 
 
+def replace_penultimate_layer(model):
+    penultimate_layer = model.model.decoder.layers[-2]
+    penultimate_layer.fc1 = QuantizedLinearLayer(penultimate_layer.fc1)
+    penultimate_layer.fc2 = QuantizedLinearLayer(penultimate_layer.fc2)
```
Comment (BichengYing) [inline cezo_fl_main.py:106]: 1. you can put this util func into `quantized_layer.py`.
2. Add type annotation and docstring
3. I feel you need to add a unit test to show the `QuantizedLinearLayer` is indeed applied

File: shared/quantized_layer.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import torch
+
+
+class QuantizedLinearLayer(torch.nn.Module):
```
Comment (BichengYing) [inline shared/quantized_layer.py:4]: please add docstring

**Review by BichengYing — CHANGES_REQUESTED** [review]
Remember to add the PR description as well

File: cezo_fl_main.py (hunk 1 above)
Comment (RogerRogerUSC) [reply cezo_fl_main.py:74]: I will create a new independent file for this.

File: zo_compression_main.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        replace_layer(model)
+        model.model_name = large_model
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_name, padding_side="left", truncate_side="left"
+        )
+        template = LM_TEMPLATE_MAP[args.dataset]()
+        verbalizer_id_map = template.get_verbalizer_id(tokenizer)
+        criterion = get_lm_loss("last_token", verbalizer_id_map)
+        optimizer = torch.optim.SGD(model.parameters(), lr=args.lr, momentum=0, weight_decay=5e-4)
+        accuracy_func = get_lm_loss("accuracy", verbalizer_id_map)
```
Comment (BichengYing) [inline zo_compression_main.py:84]: just keep what is supported and delete the rest dataset.

File: zo_compression_main.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+from models.cnn_mnist import CNN_MNIST
+from models.lenet import LeNet
+from models.cnn_fashion import CNN_FMNIST
+from models.lstm import CharLSTM
+from shared.language_utils import get_lm_loss, LM_TEMPLATE_MAP, SUPPORTED_LLM
+from shared.metrics import accuracy
+
+from tqdm import tqdm
+from gradient_estimators.random_gradient_estimator import RandomGradientEstimator as RGE
+from shared.quantized_layer import replace_layer
```
Comment (BichengYing) [inline zo_compression_main.py:24]: should import module only instead of func directly

File: zo_compression_main.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+
+def get_size_of_model(model):
+    return sum(p.numel() * p.element_size() for p in model.parameters())
+
+
+if __name__ == "__main__":
+    args = get_params().parse_args()
+    if args.dataset == "shakespeare":
+        args.num_clients = 139
```
Comment (BichengYing) [inline zo_compression_main.py:173]: delete here as well

File: zo_compression_main.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        )
+        writer = SummaryWriter(
+            path.join(
+                "tensorboards",
+                "cezo_fl",
+                args.dataset,
+                args.log_to_tensorboard,
+                tensorboard_sub_folder,
+            )
+        )
```
Comment (BichengYing) [inline zo_compression_main.py:191]: do you want to make it write into different folder?

File: shared/quantized_layer.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    def forward(self, x):
+        x = self.quantized_layer(x)
+        x = self.original_layer(x)
+        x = self.dequantized_layer(x)
+        return x
+
+
+# Quantize the full connection layers
+def replace_layer(model):
```
Comment (BichengYing) [inline shared/quantized_layer.py:19]: function docstring is
```python
def replace_layer(model):
    """func docstring""""
```
see https://google.github.io/styleguide/pyguide.html#383-functions-and-methods
Also, you missed the type annotation

**Review by BichengYing — CHANGES_REQUESTED** [review]

**Comment by ZidongLiu** [comment]
@RogerRogerUSC is this one ready for review?

File: zo_compression_main.py (hunk 6 above)
Comment (RogerRogerUSC) [reply zo_compression_main.py:173]: Deleted.

File: zo_compression_main.py (hunk 5 above)
Comment (RogerRogerUSC) [reply zo_compression_main.py:24]: Fixed.

File: zo_compression_main.py (hunk 6 above)
Comment (RogerRogerUSC) [reply zo_compression_main.py:173]: If I keep the Shakespeare dataset, should I keep these codes?

File: zo_compression_main.py (hunk 7 above)
Comment (RogerRogerUSC) [reply zo_compression_main.py:191]: Yes, modified "cezo_fl" in line 186 to "zo_compression".

File: zo_compression_main.py (hunk 4 above)
Comment (BichengYing) [reply zo_compression_main.py:84]: the original code is still in the `decom_fl_main.py`. It is always easy to copy-paste back when you are able to support compression on these models. It is always more time-consuming to debug unexpected behavior than copy-paste the code.

File: zo_compression_main.py (hunk 6 above)
Comment (BichengYing) [reply zo_compression_main.py:173]: See my comment above. Still prefer deleting it.

File: shared/quantized_layer.py (hunk 3 above)
Comment (RogerRogerUSC) [reply shared/quantized_layer.py:5]: Added.

**Comment by ZidongLiu** [comment]
Close this PR because current implementation has some defects which will be improved and reworked later

**Pull request closed without merging**